app/src/main/java/org/schabi/newpipe/info_list/holder/RepliesHandler.java

```diff
@@ -0,0 +1,192 @@
+
+            @SuppressLint("SetTextI18n")
+            @Override
+            public void onSubscribe(@NonNull final Disposable d) {
+                showReplies.setText("Downloading Replies");
+            }
+
+            @Override
+            public void onSuccess(@NonNull final CommentsInfo commentsInfo) {
+                final Single<ListExtractor.InfoItemsPage<CommentsInfoItem>>
+                        getRepliesInfoSingle =
+                        repliesSingle(commentsInfo, parentInfoItem);
+
+                final SingleObserver<ListExtractor.InfoItemsPage<CommentsInfoItem>>
+                        getRepliesInfoObserver = repliesObserver(parentInfoItem);
+
+                getRepliesInfoSingle
+                        .subscribeOn(Schedulers.io())
+                        .observeOn(AndroidSchedulers.mainThread())
+                        .subscribe(getRepliesInfoObserver);
+            }
+
+            @SuppressLint("SetTextI18n")
+            @Override
+            public void onError(@NonNull final Throwable e) {
+                showReplies.setText("Error getting replies");
+            }
+        };
+    }
+
+
+    public void addRepliesToUI(final CommentsInfoItem parentInfoItem) {
+        ((InfoListAdapter) Objects.requireNonNull(repliesView.getAdapter()))
+                .setInfoItemList(cachedReplies);
+
+        final ViewGroup.MarginLayoutParams params =
+                (ViewGroup.MarginLayoutParams) repliesView.getLayoutParams();
+        params.topMargin = 45;
+
+        repliesView.setMinimumHeight(100);
+        repliesView.setHasFixedSize(true);
+        parentInfoItem.setRepliesOpen(true);
+        showReplies.setVisibility(View.GONE);
+        repliesView.setVisibility(View.VISIBLE);
+    }
+
+    public void downloadReplies(final CommentsInfoItem parentInfoItem) {
+
+        final Single<CommentsInfo> parentInfoSingle = ExtractorHelper.getCommentsInfo(
+                parentInfoItem.getServiceId(),
+                parentInfoItem.getUrl(),
+                false
+        );
+
+        final SingleObserver<CommentsInfo> singleInfoRepliesInfoObserver
+                = repliesInfoObserver(parentInfoItem);
+
+        parentInfoSingle
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(singleInfoRepliesInfoObserver);
+    }
+
+    public void addReplies(final CommentsInfoItem parentInfoItem) {
+        if (parentInfoItem.getReplies() == null) {
+            return;
+        }
+
+        if (cachedReplies.isEmpty()) {
+            downloadReplies(parentInfoItem);
+        } else {
+            addRepliesToUI(parentInfoItem);
+        }
+    }
+
+    public void checkForReplies(final CommentsInfoItem item) {
+        if (item.getReplies() == null) {
+            repliesView.setVisibility(View.GONE);
+            showReplies.setVisibility(View.GONE);
+        } else if (item.getRepliesOpen()) {

```

Comment:
As I said in the extractor, this method should be removed from there and implemented here in the frontend. The usual approach would be to create an item to use in the `RecyclerView` that wraps around `CommentsInfoItem` but also has a field `isExpanded`. This option is not viable, though, since in NewPipe lists use a common pattern to load, reload, etc., so changing the type of data fed to the recycler view without changing the type of data returned by the extractor is not doable without huge code changes (and we don't want those). A solution that would also be suitable for the multiple-depth replies implementation (see my other comment), would be to just keep a `static Set<CommentsInfoItem>` somewhere storing a common list of expanded `CommentsInfoItem`s (common meaning that it will be accessed by *all* (even nested) `RecyclerView`s' `CommentsInfoItem.updateFromItem()`) . When the user expands a comment view the related `CommentsInfoItem` should be added to the `Set`, when the user unexpands the same comment it should be removed and when the `RecyclerView` tries to load a comment using `CommentsInfoItemHolder.updateFromItem(item)` it is expanded based on whether `item` belongs to the set.

Reply:
@litetex do you have any idea whether this can be achieved in a better way, without storing any separate data in `RecyclerView`s (since that data would be lost when the comment view is recycled)?

Reply:
@Stypox At first: Welcome back.
Whoa that's a lot of info. 

> do you have any idea whethe ...

No I don't - currently.

I think I would have to look at this in more detail to get an answer to that, however some ideas from my side to make nested commets working:
* ViewModel which holds the data?
* Write a custom Custom component for this?

Reply:
@golfinq ping ;-)

Reply:
@Stypox @litetex 
>     * ViewModel which holds the data?
>     * Write a custom Custom component for this?

I am more inclined to use the first option as I am trying to keep the amount of new code minimal for future maintenance reasons. I was hoping that by reusing the 'InfoListAdapter` the comment views which were created would automatically have the reply handling code inside of it. How are the comment views which are being created here different than the ones which are created in the main comment holding view?

Reply:
> as it currently stands this code can't handle recursive levels of replies

This is what I could guess by looking at the code, but I might be wrong since I couldn't test. But the problem is also the fact that you shouldn't store data about whether the replies menu is open or not in an object coming from the extractor. The extractor shouldn't have anything to do with the UI.